frame/composable-traits/src/financial_nft.rs

```diff
@@ -0,0 +1,152 @@
+///        *nft.x = nft.x.checked_sub(amount);
+///        Ok(nft.x)
+//       })
+///   }
+/// }
+pub trait FinancialNFTProtocol {
+	/// Abstract type of an account id.
+	type AccountId: Copy + Eq;
+
+	/// Abstract type of a class id.
+	type ClassId: Encode;
+
+	/// Abstract type of an instance id. Used to uniquely identify NFTs.
+	type InstanceId;
+
+	/// NFT provider from which we load/store NFT's.
+	type NFTProvider: FinancialNFTProvider<
+		Self::AccountId,
+		ClassId = Self::ClassId,
+		InstanceId = Self::InstanceId,
+	>;
+
+	/// Mint a new NFT into an account.
+	///
+	/// Arguments
+	///
+	/// * `account_id` the owner of the minted NFT.
+	/// * `nft` the initial value of the minted NFT.
+	///
+	/// Return the NFT instance id if successfull, otherwise the underlying NFT provider error.
+	fn mint_protocol_nft<NFT>(
+		account_id: &Self::AccountId,
+		nft: &NFT,
+	) -> Result<Self::InstanceId, DispatchError>
+	where
+		NFT: Get<Self::ClassId> + Encode,
+	{
+		let nft_class = NFT::get();
+		Self::NFTProvider::mint_nft(&nft_class, &account_id, &nft_class, &nft)
+	}
+
+	/// Ensure that the owner of the identifier NFT is `account_id`.
+	///
+	/// Arguments
+	///
+	/// * `account_id` the account id that should own the NFT.
+	/// * `instance_id` the NFT instance id.
+	///
+	/// Returns `Ok(())` if `account_id` is the owner of the NFT identified by `instance_id`.
+	fn ensure_protocol_nft_owner<NFT>(
+		account_id: &Self::AccountId,
+		instance_id: &Self::InstanceId,
+	) -> Result<(), DispatchError>
+	where
+		NFT: Get<Self::ClassId>,
+	{
+		let nft_class = NFT::get();
+		match Self::NFTProvider::owner(&nft_class, &instance_id) {
+			Some(owner_account_id) if owner_account_id == *account_id => Ok(()),
+			_ => Err(DispatchError::BadOrigin),
+		}
+	}
+
+	/// Return an NFT identified by its instance id.
+	///
+	/// Arguments
+	///
+	/// * `instance_id` the NFT instance id.
+	fn get_protocol_nft<NFT>(instance_id: &Self::InstanceId) -> Result<NFT, DispatchError>
+	where
+		NFT: Get<Self::ClassId> + Decode,
+	{
+		let nft_class = NFT::get();
+		Self::NFTProvider::typed_attribute(&nft_class, &instance_id, &nft_class)
+			.ok_or(DispatchError::Token(TokenError::UnknownAsset))
+	}
+
+	/// Mutate the NFT identified by `instance_id`.
+	///
+	/// Arguments
+	///
+	/// * `T` the callback return value.
+	/// * `E` the callback error value.
+	///
+	/// * `instance_id` the NFT instance id.
+	/// * `f` the callback in charge of mutating the NFT.
+	///
+	/// Returns the result of the callback, either `T` or `E`.
+	fn try_mutate_protocol_nft<NFT, T, E>(
+		instance_id: &Self::InstanceId,
+		f: impl FnOnce(&mut NFT) -> Result<T, E>,
+	) -> Result<T, E>
+	where
+		NFT: Get<Self::ClassId> + Encode + Decode,
+		E: From<DispatchError>,
+	{
+		let mut nft = Self::get_protocol_nft(&instance_id)?;
+		let r = f(&mut nft)?;
+		let nft_class = NFT::get();
+		Self::NFTProvider::set_typed_attribute(&nft_class, &instance_id, &nft_class, &nft)?;
```

Comment:
Already done using a `Version` associated to the NFT type. This will allow us to migrate by incrementing the version (used as attribute key). This code is outdated